src/digest.c

```diff
@@ -72,16 +72,78 @@ FILE* open_with_widechar_on_windows(const char* txt) {
     return out;
 }
 
+
+// Also already used in sha2.h
+//
+// We can rely on WORDS_BIGENDIAN only be defined on big endian systems thanks to Rconfig.
+//
+// A number of other #define based tests are in other source files here for different hash
+// algorithm implementations notably crc32c, pmurhash, sha2 and xxhash
+//
+// A small and elegant test is also in package qs based on https://stackoverflow.com/a/1001373
+
+// edd 02 Dec 2013  use Rconfig.h to define BYTE_ORDER, unless already defined
+#ifndef BYTE_ORDER
+// see sha2.c comments, and on the internet at large
+#define LITTLE_ENDIAN 1234
+#define BIG_ENDIAN    4321
+#ifdef WORDS_BIGENDIAN
+#define BYTE_ORDER  BIG_ENDIAN
+#else
+#define BYTE_ORDER  LITTLE_ENDIAN
+#endif
+#endif
+
+SEXP is_big_endian(void) {
+    return Rf_ScalarLogical(BYTE_ORDER == BIG_ENDIAN);
+}
+
+SEXP is_little_endian(void) {
+    return Rf_ScalarLogical(BYTE_ORDER == LITTLE_ENDIAN);
+}
+
+/* for any algorithm that has output as unsigned char*, can use resolve */
+#define resolve(WHAT) if (leaveRaw) {                          \
+    memcpy(output, (WHAT), output_length);                     \
+} else for (int j = 0; j < output_length; j++) {               \
+    snprintf(output + j * 2, 3, "%02x", (WHAT)[j]);            \
+}                                                              \
+
+#define decl_tmp32(WHAT) uint32_t tmp = (WHAT);
+#define decl_tmp64(WHAT) uint64_t tmp = (WHAT);
+
+/* for any algorithm that has output as an integral value, can use dump */
+
+void rev_memcpy(char *dst, const void *src, int len) {
+    for (int i = 0; i < len; i++) {
+        dst[i] = ((char*)src)[len - i - 1];
+    }
+}
+
+#define dump32(WHAT) decl_tmp32(WHAT);            \
```

Comment:
Can we use `_dump32` or something more evocative?

Reply:
propose `_store_from_int32` (and `_store_from_int64` re next comment)